File: src/common/net/auoeke/wheel/dependency/WheelDependencyFactory.java
package net.auoeke.wheel.dependency;

import net.auoeke.wheel.extension.WheelExtension;
import net.auoeke.wheel.extension.dependency.Dependency;
import java.util.Map;
import org.gradle.api.artifacts.dsl.RepositoryHandler;
import org.gradle.api.artifacts.repositories.ArtifactRepository;
import org.gradle.api.artifacts.repositories.MavenArtifactRepository;
import org.gradle.api.internal.artifacts.DefaultDependencyFactory;
import user11681.reflect.Classes;
import net.auoeke.wheel.WheelPlugin;

@SuppressWarnings("ConstantConditions")
public class WheelDependencyFactory extends DefaultDependencyFactory {
    public static final long klass = Classes.klass(WheelDependencyFactory.class);

    public WheelDependencyFactory() {
        super(null, null, null, null, null, null);
    }

    private static String changeVersion(String artifact, String version) {
        String[] segments = artifact.split(":");
        segments[2] = version;

        return String.join(":", segments);
    }

    private static void addRepository(String repository) {
        if (WheelPlugin.currentProject != null && repository != null) {
            RepositoryHandler repositories = WheelPlugin.currentProject.getRepositories();

            for (ArtifactRepository artifactRepository : repositories) {
                if (artifactRepository instanceof MavenArtifactRepository && repository.equals(((MavenArtifactRepository) artifactRepository).getUrl().toString())) {
                    return;
                }
            }

            repositories.maven((MavenArtifactRepository artifactRepository) -> artifactRepository.setUrl(repository));
        }
    }

    private static boolean addRepository(Dependency entry) {
        if (entry != null) {
            addRepository(entry.resolveRepository());

            return true;
        }

        return false;
    }

    private static Object resolve(String dependency) {
        String[] components = dependency.split(":");

        if (components.length == 2) {
            Dependency entry = WheelExtension.dependency(components[0]);

            if (addRepository(entry)) {
                return changeVersion(entry.artifact, components[1]);
            }
        }

        Dependency entry = WheelExtension.dependency(dependency);

        if (addRepository(entry)) {
            return entry.artifact;
        }

        if (WheelPlugin.currentProject != null && WheelPlugin.currentProject.findProject(dependency) != null) {
            return WheelPlugin.currentProject.getDependencies().project(Map.of("path", dependency));
        }

        return dependency;
    }

    @Override
    public org.gradle.api.artifacts.Dependency createDependency(Object dependencyNotation) {
        return dependencyNotation instanceof String ? super.createDependency(resolve((String) dependencyNotation)) : super.createDependency(dependencyNotation);
    }
}

File: src/net/auoeke/wheel/dependency/WheelDependencyFactory.kt
package net.auoeke.wheel.dependency

import net.auoeke.extensions.string
import net.auoeke.extensions.then
import net.auoeke.extensions.type
import net.auoeke.reflect.Classes
import net.auoeke.wheel.extension.WheelExtension
import org.gradle.api.Project
import org.gradle.api.artifacts.Dependency
import org.gradle.api.artifacts.repositories.MavenArtifactRepository
import org.gradle.api.internal.artifacts.DefaultDependencyFactory
import net.auoeke.wheel.extension.dependency.Dependency as WheelDependency

@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
class WheelDependencyFactory : DefaultDependencyFactory(null, null, null, null, null, null) {
    override fun createDependency(dependencyNotation: Any): Dependency = super.createDependency(when (dependencyNotation) {
        is String -> this.resolve(dependencyNotation)
        else -> dependencyNotation
    })

    private fun withVersion(artifact: String, version: String): String = (artifact.split(":") as MutableList).also {
        when {
            it.size > 2 -> it[2] = version
            else -> it += version
        }
    }.joinToString(":")

    private fun addRepository(repository: String?) {
        if (project !== null && repository !== null) {
            val repositories = project!!.repositories

            for (artifactRepository in repositories) {
                if (artifactRepository is MavenArtifactRepository && repository == artifactRepository.url.string) {
                    return
                }
            }

            repositories.maven {it.setUrl(repository)}
        }
    }

    private fun addRepository(entry: WheelDependency?): Boolean = (entry !== null).then {
        this.addRepository(entry!!.resolveRepository())
    }

    private fun resolve(dependency: String): Any {
        val components = dependency.split(":") as MutableList

        if (components.size == 2) {
            val entry = WheelExtension.dependency(components[0])

            return when (this.addRepository(entry)) {
                true -> this.withVersion(entry!!.artifact, components[1])
                false -> this.withVersion(dependency, "latest.release")
            }
        }

        val entry = WheelExtension.dependency(dependency)

        return when {
            this.addRepository(entry) -> entry!!.artifact
            project?.findProject(dependency) !== null -> project!!.dependencies.project(mapOf("path" to dependency))
            else -> dependency
        }
    }

    companion object {
        val klass = Classes.klass(type<WheelDependencyFactory>())
        var project: Project? = null
    }
}
